gotsys/gotsys.conf.5
.\"
.\" Copyright (c) 2024 Stefan Sperling <stsp@openbsd.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate$
.Dt GOTSYS.CONF 5
.Os
.Sh NAME
.Nm gotsys.conf
.Nd gotsys configuration file
.Sh DESCRIPTION
.Nm
provides system configuration information to
.Xr gotsysd 8 .
.Pp
.Nm
allows remote administrators to configure aspects of Git repositories
hosted by
.Xr gotd 8
without having shell access to the Git server's operating system.
Remote administrators merely need to be granted write access to a
special-purpose
.Em system repository
called
.Pa gotsys.git
in order to configure Git repository services.
.Pp
The file format is line-based, with one configuration directive per line.
Any lines beginning with a
.Sq #
are treated as comments and ignored.
.Sh GLOBAL CONFIGURATION
 The available global configuration directives are as follows:
.Bl -tag -width Ds
.It Ic group Ar name
Declare a group with the given
.Ar name .
Users can then be granted membership of this group in order to manage
repository access rules on a per-group basis.
.Pp
Group names may only contain alphabetic ASCII characters (a-z, A-Z),
non-leading digits (0-9), non-leading hyphens (-),
non-leading underscores (_), and non-leading periods (.).
.Pp
The
.Ar name
.Dq anonymous
is reserved for use with the
.Ic permit
directive and cannot be used with the
.Ic group
directive.
.Pp
User and group names which are used for special purposes by the
.Ox
system cannot be used in
.Nm .
The current list of reserved names is
.Dq build ,
.Dq root ,
.Dq wheel ,
.Dq daemon ,
.Dq kmem ,
.Dq sys ,
.Dq tty ,
.Dq operator ,
.Dq bin ,
.Dq wsrc ,
.Dq users ,
.Dq auth ,
.Dq games ,
.Dq staff ,
.Dq wobj ,
.Dq sshd ,
.Dq guest ,
.Dq utmp ,
.Dq crontab ,
.Dq www ,
.Dq network ,
.Dq authpf ,
.Dq dialer ,
.Dq nogroup ,
and
.Dq nobody .
.It Ic user Ar name
Declare a user which can then be granted access to specific repositories.
.Pp
A user must be declared with a unique
.Ar name ,
followed by user-specific configuration directives inside curly braces:
.Pp
.Ic user Ar name Brq ...
.Pp
User names may only contain alphabetic ASCII characters (a-z, A-Z),
non-leading digits (0-9), non-leading hyphens (-),
non-leading underscores (_), and non-leading periods (.).
.Pp
The
.Ar name
.Dq anonymous
is reserved for use with the
.Ic permit
directive and cannot be used with the
.Ic user
directive.
.Pp
User and group names used for special purposes by the
.Ox
system cannot be used in
.Nm .
The list of reserved user names is the same as documented above for the
.Ic group
directive.
.Pp
When a user's declaration is removed from
.Nm
the server account password will be locked, and any previously installed
authorized keys will be removed, rendering the account inaccessible.
The account can be made accessible again by restoring the user's
.Nm
entry.
.Pp
The available
.Cm user
parameters are as follows:
.Bl -tag -width Ds
.It Ic group Ar group
Make the user a member of the given
.Ar group .
This directive may be specified multiple times to add the user to
multiple groups.
All groups listed here must already have been declared with the
.Ic group
directive.
.It Ic password Ar string
Set a password for the user.
The
.Ar string
argument must be an encrypted password string, as generated by OpenBSD's
.Xr encrypt 1
command.
If not specified, password-based authentication will be disabled for this user.
.It Ic authorized Ic key Ar type Ar key Oo Ar comment Oc
Allow the user to authenticate with an SSH public key.
.Pp
The list of arguments consist of the following space-separated fields:
the key
.Ar type ,
the base64-encoded 
.Ar key ,
and an optional
.Ar comment .
These arguments will usually correspond to the contents of an
.Pa id_ecdsa.pub ,
.Pa id_ecdsa_sk.pub ,
.Pa id_ed25519.pub ,
.Pa id_ed25519_sk.pub ,
or
.Pa id_rsa.pub
file provided by the user.
.Pp
.Pp
The supported key types are:
.Pp
.Bl -item -compact -offset indent
.It
sk-ecdsa-sha2-nistp256@openssh.com
.It
ecdsa-sha2-nistp256
.It
ecdsa-sha2-nistp384
.It
ecdsa-sha2-nistp521
.It
sk-ssh-ed25519@openssh.com
.It
ssh-ed25519
.It
ssh-rsa
.El
.Pp
The key type must be followed by the base64-encoded public
.Ar key .
.Pp
The optional
.Ar comment
is not used for anything, but may be
convenient for the user to identify the key.
.Pp
The
.Ic authorized Ic key
directive may be specified multiple times to allow the user to
authenticate with any of the specified keys.
.Pp
Two-factor authentication with FIDO keys can be enforced by only
adding keys of type
.Em sk-ssh-ed25519@openssh.com
or
.Em sk-ecdsa-sha2-nistp256@openssh.com .
.El
.\" Set the path to the unix socket which
.\" .Xr gotd 8
.\" should listen on.
.\" If not specified, the path
.\" .Pa /var/run/gotd.sock
.\" will be used.
.\" .It Ic user Ar user
.\" Set the
.\" .Ar user
.\" which will run
.\" .Xr gotd 8 .
.\" Initially,
.\" .Xr gotd 8
.\" requires root privileges in order to create its unix socket.
.\" Afterwards,
.\" .Xr gotd 8
.\" drops privileges to the specified
.\" .Ar user .
.\" If not specified, the user _gotd will be used.
.El
.Sh REPOSITORY CONFIGURATION
Repositories declared in
.Nm
will be created on the server if they do not already exist.
Each repository must be given a unique
.Ar name ,
followed by repository-specific configuration directives inside curly braces:
.Pp
.Ic repository Ar name Brq ...
.Pp
Repository names may only contain alphabetic ASCII characters (a-z, A-Z),
non-leading digits (0-9), non-leading hyphens (-),
non-leading underscores (_), and non-leading periods (.).
.Pp
For each repository, access rules must be configured using the
.Ic permit
and
.Ic deny
configuration directives.
Multiple access rules can be specified, and the last matching rule
determines the action taken.
If no rule matches, access to the repository is denied.
.Pp
Reference and branch names specified in
.Nm
are non-empty UTF-8 strings with the following disallowed characters:
.Sq \ \&
(space, tab, formfeed, newline, carriage return),
\(a~ (tilde),
\(a^ (caret),
: (colon),
? (question mark),
* (asterisk),
[ (opening square bracket),
\\ (backslash),
" (double quote),
\' (single quote),
{ (opening curly brace),
} (closing curly brace),
= (equal sign),
# (hash),
$ (dollar sign).
The name may not be the string
.Dq @ ,
may not begin with a / (slash), and may not begin or end
with a \. (period).
Additionally, the name may not contain the two-character sequences
//, .. , and @{.
.Pp
The available repository configuration directives are as follows:
.Bl -tag -width Ds
.It Ic description Ar string
Sets the repository description shown on the
.Xr gotwebd 8
repository listing page.
The
.Ar string
parameter can only contain one line of text in ASCII encoding,
or in UTF-8 encoding if
.Ar string
is wrapped in quotes.
.It Ic hash algorithm Ar hashing-algorithm
Configure the repository's
.Ar hashing-algorithm
used for the computation of Git object IDs.
Possible values are
.Dq sha1
.Pq the default
or
.Dq sha256 .
.Pp
This configuration directive only affects newly created repositories.
The hashing algorithm of existing repositories cannot be changed.
.It Ic head Ar branch
Point the repository's symbolic
.Pa HEAD
reference at the specified
.Ar branch .
If not specified,
.Pa HEAD
will point at the branch
.Dq main ,
regardless of whether this branch actually exists in the repository.
.Pp
If
.Pa HEAD
points at a non-existent branch then clients may fail to clone the repository
because they rely on
.Pa HEAD
to determine which branch to fetch by default.
.It Ic deny Ar identity
Deny repository access to users with the username
.Ar identity .
Group names may be matched by prepending a colon
.Pq Sq \&:
to
.Ar identity .
.It Ic permit Ar mode Ar identity
Permit repository access to users with the username
.Ar identity .
The
.Ar mode
argument must be set to either
.Ic ro
for read-only access,
or
.Ic rw
for read-write access.
Group names may be matched by prepending a colon
.Pq Sq \&:
to
.Ar identity .
.Pp
The special user
.Ar identity
.Dq anonymous
can be used when public read-only access to repositories over SSH is desired.
The anonymous user has an empty password, cannot use an SSH public key, and
can only be granted read-only access.
.It Ic protect Brq Ar ...
The
.Cm protect
directive may be used to protect branches and tags in a repository
from being overwritten by potentially destructive client-side commands,
such as when
.Cm got send -f
and
.Cm git push -f
are used to change the history of a branch.
.Pp
To build a set of protected branches and tags, multiple
.Ic protect
directives may be specified per repository and
multiple
.Ic protect
directive parameters may be specified within curly braces.
.Pp
The available
.Cm protect
parameters are as follows:
.Bl -tag -width Ds
.It Ic branch Ar name
Protect the named branch.
The branch may be created if it does not exist yet.
Attempts to delete the branch or change its history will be denied.
.Pp
If the
.Ar name
does not already begin with
.Dq refs/heads/
it will be looked up in the
.Dq refs/heads/
reference namespace.
.It Ic branch Ic namespace Ar namespace
Protect the given reference namespace, assuming that references in
this namespace represent branches.
New branches may be created in the namespace.
Attempts to change the history of branches or delete them will be denied.
.Pp
The
.Ar namespace
argument must be absolute, starting with
.Dq refs/ .
.It Ic tag Ic namespace Ar namespace
Protect the given reference namespace, assuming that references in
this namespace represent tags.
New tags may be created in the namespace.
Attempts to change or delete existing tags will be denied.
.Pp
The
.Ar namespace
argument must be absolute, starting with
.Dq refs/ .
.El
.Pp
The special reference namespaces
.Dq refs/got/
and
.Dq refs/remotes/
do not need to be listed in
.Nm .
These namespaces are always protected and even attempts to create new
references in these namespaces will always be denied.
.It Ic notify Brq Ar ...
The
.Ic notify
directive enables notifications about new commits or tags
added to the repository.
.Pp
The default content of email notifications looks similar to the output of the
.Cm got log -d
command.
.Pp
Notifications via HTTP require a HTTP or HTTPS server which is accepting
POST requests with or without HTTP Basic authentication.
Depending on the use case a custom server-side CGI script may be required
for the processing of notifications.
HTTP notifications can achieve functionality
similar to Git's server-side post-receive hook script
by triggering arbitrary post-commit actions via the HTTP server.
.Pp
The

CHANGES
* got 0.129; 2026-09-26
- fix missing status code of staged files in log message template during commit
- remove empty dangling directories within refs/ when removing references
- do not add unversioned files to version control if a file addition is merged
- handle removal of UID_MAX and GID_MAX
- fix a problem where 'got revert' errors out in work trees with a path-prefix
- plug a memory leak caused by duplicate ignore patterns

* got 0.128; 2026-09-04
- switch from imsg_get(), removed in OpenBSD-current, to imsgbuf_get() 
- fix tp_urlescape() escaping in template tool for gotwebd
- fix off-by-one in pack index verification: check up to fanout_table[0xff]
- fix ignore pattern matching when a file path is given to 'got status'
- plug several memory leaks reachable via 'got histedit' and other commands
- fix bogus "bad object" errors from 'got merge -a' when merging a tag
- fix segfault in error handling path of gotsys-apply-conf
- fix bogus "no such file or directory" errors seen during histedit

* got 0.127; 2026-07-20
- re-add cmdline tests to 'make regress' target (regression from 0.126)
- add version flag (-V, --version) to gotd, gotwebd, and gotsysd 
- make 'got merge' accept tags as merge source argument
- do not raise errors when unversioned files are found during got rm -k
- add rm -s ? option for deleting unversioned files from the work tree
- add rm -I option for also deleting ignored unversioned files with rm -s ?
- make 'tog log' reload the work tree's base-commit info during Ctrl-L refresh
- fix failures due to trailing slashes in checkout -p path-prefix argument
- gotwebd: use <meta http-equiv="refresh"> tag to redirect browser after login
- gotwebd.conf: add option which restores legacy redirect using http status 307

* got 0.126; 2026-05-21
- remove dependency on xxd from the test suite (dep was added in got-0.125)
- really add all relevant parent commits to newly created pack files
- fix serving a branch of entirely unrelated history with gotd
- fix test failures when tests are run with Git 2.54
- initial sha256 support in the network protocol for got clone/fetch/send

* got 0.125; 2026-05-10
- security fix: reject versioned files inside .git, .got, or .cvg directories
- security fix: crafted tree entry names could cause writes outside work tree
- fix redundant pack file cleanup when repository contains symlinks
- prevent NULL pointer dereferences when empty tree objects are encountered

* got 0.124; 2026-04-13
- make the chroot path directive in gotwebd.conf actually work
- fix a segfault in tog while using the & search feature
- plug a tree object leak in the gotd repo_write process
- fix gotd wrongly complaining about a missing gotsys.conf in pack files
- expand tabs in log messages displayed by tog diff to prevent misalignment
- prevent non-root users from blocking gotctl reload requests
- plug a memory leak in got-read-commit
- allow UTF-8 in gotsys.conf site owner names and repository descriptions
- reject non-UTF-8-encoded reference names in gotsys.conf
- make gotwebd display logged-in usernames in case of group-membership auth

* got 0.123; 2026-02-25
- make gotsys-write-conf configure clone-urls for all accessible repositories
- ensure visitors see the repository index page after logging into gotwebd
- make 'gotadmin cleanup' run even if HEAD points at a non-existent branch
- gotsys.conf.5 and got.1 wording and markup fixes
- replace obsolete tmppath pledge in got-notify-http with wpath+cpath & unveil
- avoid a malloc/free dance per parsed tree entry in got-read-pack
- stop using the pack delta-cache in got-read-pack, cache-less is faster here
- fix double-free in error path of the 'gotadmin pack' commit coloring phase
- store first-level object_idset hash table entries inline to avoid malloc/free
- avoid doing asprintf/free per tree entry in got_pack_load_tree_entries()
- avoid a per tree-entry memcpy() in got-read-pack enumerate_tree()
- avoid deltifying packed delta-base objects to speed up pack file generation
- cache fewer but larger deltas in delta-cache to speed up got-index-pack

* got 0.122; 2026-02-10
- fix gotd/gotsys-check trying to parse gotsys.conf from a non-zero file offset
- fix setting of a custom HEAD reference during repository creation by gotsysd
- tweak gotwebd's Content-Security-Policy (CSP) to allow javascript in websites
- fix CSP-violating URL-paths to gotwebd static assets in generated HTML
- document gotwebd's Content-Security-Policy in gotwebd.conf.5
- fix gotwebd serving websites from a branch other than HEAD
- add a per-repository clone_url option to gotwebd.conf
- make gotsys-write-conf set the clone_url option in gotwebd.conf
- make 'tog diff' indent log messages above patches for safety
- gotweb can now show SSH host key fingerprints for clone URLs and login hints
- make gotsys-write-conf add ssh host key fingerprints to /etc/gotwebd.conf
- only display the "Clone URL:" label if there is a clone URL to display
- fix overflow CSS for gotwebd clone URL
- scope gotwebd authentication cookies to gotweb URL root path instead of "/"
- make gotwebd display the authenticated user name while logged in
- provide a link which can be used to log out when logged into gotwebd

* got 0.121; 2026-01-31
- plug fulltext delta cache memory leaks; found by otto malloc
- make gotwebd serve its own static files if not they are not served by httpd
- avoid deleting local HEAD and "refs/got/" refs during 'got fetch -d'
- hint at PASSWDSKIP for security(8) in gotsh.1
- fix bogus 'got histedit' error when deleting files which do not exist on disk

gotsys/gotsys.c
/*
 * Copyright (c) 2025 Stefan Sperling <stsp@openbsd.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/queue.h>
#include <sys/tree.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/stat.h>

#include <err.h>
#include <errno.h>
#include <event.h>
#include <fcntl.h>
#include <imsg.h>
#include <sha1.h>
#include <sha2.h>
#include <limits.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <unistd.h>

#include "got_error.h"
#include "got_version.h"
#include "got_path.h"
#include "got_opentemp.h"
#include "got_repository.h"
#include "got_reference.h"
#include "got_object.h"

#include "media.h"
#include "gotwebd.h"
#include "gotsys.h"
#include "gotsysd.h"

#ifndef nitems
#define nitems(_a)	(sizeof((_a)) / sizeof((_a)[0]))
#endif

struct gotsys_cmd {
	const char	*cmd_name;
	const struct got_error *(*cmd_main)(int, char *[]);
	void		(*cmd_usage)(int);
};

__dead static void	usage(int, int);

__dead static void	usage_apply(int);
__dead static void	usage_check(int);

static const struct got_error*		cmd_apply(int, char *[]);
static const struct got_error*		cmd_check(int, char *[]);

static const struct gotsys_cmd gotsys_commands[] = {
	{ "apply",	cmd_apply,	usage_apply },
	{ "check",	cmd_check,	usage_check },
};

static const struct got_error *
unveil_repo(const char *repo_path)
{
#ifdef PROFILE
	if (unveil("gmon.out", "rwc") != 0)
		return got_error_from_errno2("unveil", "gmon.out");
#endif
	if (unveil(repo_path, "r") != 0)
		return got_error_from_errno2("unveil", repo_path);

	if (unveil(GOT_TMPDIR_STR, "rwc") != 0)
		return got_error_from_errno2("unveil", GOT_TMPDIR_STR);

	if (unveil(NULL, NULL) != 0)
		return got_error_from_errno("unveil");

	return NULL;
}

static const struct got_error *
cat_blob(struct got_object_id *id, struct got_repository *repo, int fd,
    FILE *outfile)
{
	const struct got_error *err;
	struct got_blob_object *blob;

	err = got_object_open_as_blob(&blob, repo, id, 8192, fd);
	if (err)
		goto done;

	err = got_object_blob_dump_to_file(NULL, NULL, NULL, outfile, blob);
done:
	if (blob)
		got_object_blob_close(blob);
	return err;
}

static const struct got_error *
read_gotsysconf(struct got_object_id **commit_id, struct got_repository *repo,
    const char *commit_id_str, const char *filename, int tmpfd, FILE *outfile)
{
	const struct got_error *err = NULL;
	struct got_reflist_head refs;
	struct got_object_id *id = NULL;
	struct got_commit_object *commit = NULL;
	int obj_type;

	TAILQ_INIT(&refs);

	*commit_id = NULL;

	err = got_ref_list(&refs, repo, NULL, got_ref_cmp_by_name, NULL);
	if (err)
		goto done;

	err = got_repo_match_object_id(commit_id, NULL,
	    commit_id_str, GOT_OBJ_TYPE_COMMIT, &refs, repo);
	if (err)
		goto done;

	err = got_object_open_as_commit(&commit, repo, *commit_id);
	if (err)
		goto done;

	err = got_object_id_by_path(&id, repo, commit, filename);
	if (err)
		goto done;

	err = got_object_get_type(&obj_type, repo, id);
	if (err)
		goto done;

	if (obj_type != GOT_OBJ_TYPE_BLOB) {
		err = got_error_path(filename, GOT_ERR_OBJ_TYPE);
		goto done;
	}

	err = cat_blob(id, repo, tmpfd, outfile);
done:
	if (commit)
		got_object_commit_close(commit);
	got_ref_list_free(&refs);
	if (err) {
		free(*commit_id);
		*commit_id = NULL;
	}
	free(id);
	return err;
}


__dead static void
usage_apply(int status)
{
	FILE *fp = (status == 0) ? stdout : stderr;
	fprintf(fp, "usage: %s apply [-f socket] [-r repository] "
	    "[-c commit] [-w] [filename]\n", getprogname());
	exit(status);
}

static const struct got_error *
connect_gotsysd(int *gotsysd_sock, const char *socket_path)
{
	const struct got_error *err = NULL;
	struct sockaddr_un sun;

	*gotsysd_sock = -1;

	if ((*gotsysd_sock = socket(AF_UNIX, SOCK_STREAM, 0)) == -1)
		return got_error_from_errno("socket");

	memset(&sun, 0, sizeof(sun));
	sun.sun_family = AF_UNIX;
	if (strlcpy(sun.sun_path, socket_path, sizeof(sun.sun_path)) >=
	    sizeof(sun.sun_path)) {
		err = got_error_msg(GOT_ERR_NO_SPACE,
		    "gotsysd socket path too long");
		goto done;
	}

	if (connect(*gotsysd_sock, (struct sockaddr *)&sun,
	    sizeof(sun)) == -1) {
		err = got_error_from_errno2("connect", socket_path);
		goto done;
	}
done:
	if (err) {
		close(*gotsysd_sock);
		*gotsysd_sock = -1;
	}
	return err;
}

static const struct got_error *
recv_error(struct imsg *imsg)
{
	struct gotsysd_imsg_error ierr;
	size_t datalen;

	datalen = imsg->hdr.len - IMSG_HEADER_SIZE;
	if (datalen != sizeof(ierr))
		return got_error(GOT_ERR_PRIVSEP_LEN);
	memcpy(&ierr, imsg->data, sizeof(ierr));

	if (ierr.code == GOT_ERR_ERRNO)
		errno = ierr.errno_code;

	return got_error_msg(ierr.code, ierr.msg);
}

static const struct got_error *
cmd_apply(int argc, char *argv[])
{
	const struct got_error *err;
	struct imsgbuf ibuf;
	struct imsg imsg;
	struct got_repository *repo = NULL;
	char *repo_path = NULL;
	const char *commit_id_str = GOT_REF_HEAD;
	const char *filename = GOTSYSD_SYSCONF_FILENAME;
	const char *socket_path = GOTSYSD_UNIX_SOCKET;
	struct got_commit_object *commit = NULL;
	int ch, ret, fd = -1, sysconf_fd = -1, gotsysd_sock = -1, wait = 0;
	FILE *sysconf_file = NULL;
	struct got_object_id *commit_id = NULL;
	struct gotsysd_imsg_cmd_sysconf sysconf_cmd;
	int *pack_fds = NULL;
	ssize_t n;

	memset(&ibuf, 0, sizeof(ibuf));

	while ((ch = getopt(argc, argv, "f:c:r:w")) != -1) {
		switch (ch) {
		case 'c':
			commit_id_str = optarg;
			break;
		case 'f':
			socket_path = optarg;
			break;
		case 'r':
			repo_path = realpath(optarg, NULL);
			if (repo_path == NULL) {
				err = got_error_from_errno2("realpath",
				    optarg);
				goto done;
			}
			got_path_strip_trailing_slashes(repo_path);
			break;
		case 'w':
			wait = 1;
			break;
		default:
			usage_apply(1);
			/* NOTREACHED */
			break;
		}
	}

	argc -= optind;
	argv += optind;

	if (argc > 1)
		usage_apply(1);

	filename = (argc == 1 ? argv[0] : GOTSYSD_SYSCONF_FILENAME);

#ifndef PROFILE
	if (pledge("stdio rpath wpath cpath sendfd unix unveil", NULL) == -1) {
		err = got_error_from_errno("pledge");
		goto done;
	}
#endif
	err = connect_gotsysd(&gotsysd_sock, socket_path);
	if (err)
		goto done;

#ifndef PROFILE
	if (pledge("stdio rpath wpath cpath sendfd unveil", NULL) == -1) {
		err = got_error_from_errno("pledge");
		goto done;
	}
#endif
	fd = got_opentempfd();
	if (fd == -1) {
		err = got_error_from_errno("got_opentempfd");
		goto done;
	}

	sysconf_file = got_opentemp();
	if (sysconf_file == NULL) {
		err = got_error_from_errno("got_opentemp");
		goto done;
	}

	err = got_repo_pack_fds_open(&pack_fds);
	if (err != NULL)
		goto done;
#ifndef PROFILE
	if (pledge("stdio rpath sendfd unveil", NULL) == -1) {
		err = got_error_from_errno("pledge");
		goto done;
	}
#endif
	if (repo_path == NULL) {
		repo_path = strdup(GOTSYSD_REPOSITORIES_PATH "/"
		    GOTSYS_SYSTEM_REPOSITORY_NAME ".git");
		if (repo_path == NULL) {
			err = got_error_from_errno("strdup");
			goto done;
		}
	}

	err = got_repo_open(&repo, repo_path, NULL, pack_fds);
	if (err != NULL)
		goto done;

	err = unveil_repo(got_repo_get_path(repo));
	if (err)
		goto done;

	err = read_gotsysconf(&commit_id, repo, commit_id_str, filename, fd,
	    sysconf_file);
	if (err)
		goto done;
#ifndef PROFILE
	if (pledge("stdio sendfd", NULL) == -1) {
		err = got_error_from_errno("pledge");
		goto done;
	}
#endif
	if (imsgbuf_init(&ibuf, gotsysd_sock) == -1) {
		err = got_error_from_errno("imsgbuf_init");
		goto done;
	}
	imsgbuf_allow_fdpass(&ibuf);

	sysconf_fd = dup(fileno(sysconf_file));
	if (sysconf_fd == -1) {
		err = got_error_from_errno("dup");
		goto done;
	}

	memset(&sysconf_cmd, 9, sizeof(sysconf_cmd));
	memcpy(&sysconf_cmd.commit_id, commit_id,
	    sizeof(sysconf_cmd.commit_id));
	ret = imsg_compose(&ibuf, GOTSYSD_IMSG_CMD_SYSCONF, 0, getpid(),
	    sysconf_fd, &sysconf_cmd, sizeof(sysconf_cmd));
	if (ret == -1) {
		err = got_error_from_errno("imsg_compose");
		goto done;
	}
	sysconf_fd = -1;

	ret = imsgbuf_flush(&ibuf);
	if (ret == -1) {
		err = got_error_from_errno("imsgbuf_flush");
		goto done;
	}
#ifndef PROFILE
	if (pledge("stdio", NULL) == -1) {
		err = got_error_from_errno("pledge");
		goto done;
	}
#endif
	do {
		n = imsgbuf_read(&ibuf);
		if (n == -1) {
			err = got_error_from_errno("imsgbuf_read");
			goto done;
		}
		if (n == 0) {
			err = got_error(GOT_ERR_EOF);
			goto done;
		}

		n = imsgbuf_get(&ibuf, &imsg);
		if (n == -1) {
			err = got_error_from_errno("imsgbuf_get");
			goto done;
		}
		if (n == 0) {
			err = got_error(GOT_ERR_PRIVSEP_READ);
			goto done;
		}

		switch (imsg.hdr.type) {
		case GOTSYSD_IMSG_ERROR:
			err = recv_error(&imsg);
			break;
		case GOTSYSD_IMSG_SYSCONF_STARTED:
			break;
		case GOTSYSD_IMSG_SYSCONF_SUCCESS:
			printf("sysconf success\n");
			wait = 0;
			break;
		case GOTSYSD_IMSG_SYSCONF_FAILURE:
			err = got_error_fmt(GOT_ERR_ON_SERVER_SIDE,
			    "sysconf failure");
			break;
		default:
			err = got_error(GOT_ERR_PRIVSEP_MSG);
			break;
		}

		imsg_free(&imsg);
	} while (err == NULL && wait);
done:
	imsgbuf_clear(&ibuf);
	free(repo_path);
	if (commit)
		got_object_commit_close(commit);
	free(commit_id);
	if (repo) {
		const struct got_error *close_err = got_repo_close(repo);
		if (err == NULL)
			err = close_err;
	}
	if (pack_fds) {
		const struct got_error *pack_err =
		    got_repo_pack_fds_close(pack_fds);
		if (err == NULL)
			err = pack_err;
	}
	if (fd != -1 && close(fd) == -1 && err == NULL)
		err = got_error_from_errno("close");
	if (sysconf_file && fclose(sysconf_file) == EOF && err == NULL)
		err = got_error_from_errno("close");
	if (sysconf_fd != -1 && close(sysconf_fd) == -1 && err == NULL)
		err = got_error_from_errno("close");

	return err;
}

__dead static void
usage_check(int status)
{
	FILE *fp = (status == 0) ? stdout : stderr;
	fprintf(fp, "usage: %s check [-q] [-f file]\n", getprogname());
	exit(status);
}

static const struct got_error *
unveil_none(void)
{
	if (unveil("/", "") != 0)
		return got_error_from_errno("unveil");

	if (unveil(NULL, NULL) != 0)
		return got_error_from_errno("unveil");

	return NULL;
}

static const struct got_error *
unveil_conf(const char *config_file)
{
#ifdef PROFILE
	if (unveil("gmon.out", "rwc") != 0)
		return got_error_from_errno2("unveil", "gmon.out");
#endif
	if (config_file) {
		if (unveil(config_file, "r") != 0)
			return got_error_from_errno2("unveil", config_file);
	}

	if (unveil(NULL, NULL) != 0)
		return got_error_from_errno("unveil");

	return NULL;
}

static const struct got_error *
cmd_check(int argc, char *argv[])
{
	const struct got_error *err;
	int ch, fd = -1, quiet = 0;
	char *configfile = NULL;
	struct gotsys_conf gotsysconf;
	struct stat sb;

	gotsys_conf_init(&gotsysconf);

	while ((ch = getopt(argc, argv, "f:q")) != -1) {
		switch (ch) {
		case 'q':
			quiet = 1;
			break;
		case 'f':
			if (strcmp(optarg, "-") == 0) {
				fd = STDIN_FILENO;
				configfile = strdup("stdin");
				if (configfile == NULL)
					return got_error_from_errno("strdup");
				break;
			}
			configfile = realpath(optarg, NULL);
			if (configfile == NULL) {
				return got_error_from_errno2("realpath",
				    optarg);
			}
			got_path_strip_trailing_slashes(configfile);
			break;
		default:
			usage_check(1);
			/* NOTREACHED */
			break;
		}
	}
	argc -= optind;
	argv += optind;

	if (argc > 0)
		usage_check(1);

	if (fd != STDIN_FILENO && configfile == NULL) {
		configfile = strdup(GOTSYSD_SYSCONF_FILENAME);
		if (configfile == NULL)
			return got_error_from_errno("strdup");
	}

#ifndef PROFILE
	if (pledge("stdio rpath unveil", NULL) == -1) {
		err = got_error_from_errno("pledge");
		goto done;
	}
#endif
	if (fd == STDIN_FILENO) {
		err = unveil_none();
		if (err)
			goto done;
	} else {
		err = unveil_conf(configfile);
		if (err)
			goto done;

		fd = open(configfile, O_RDONLY);
		if (fd == -1) {
			err = got_error_from_errno2("open", configfile);
			goto done;
		}
	}

#ifndef PROFILE
	if (pledge("stdio", NULL) == -1) {
		err = got_error_from_errno("pledge");
		goto done;
	}
#endif
	if (fd != STDIN_FILENO) {
		if (fstat(fd, &sb) == -1) {
			err = got_error_from_errno2("fstat", configfile);
			goto done;
		}
		if (!S_ISREG(sb.st_mode)) {
			err = got_error_fmt(GOT_ERR_BAD_PATH,
			    "%s is not a regular file", configfile);
			goto done;
		}
	}

	err = gotsys_conf_parse(configfile, &gotsysconf, &fd);
	if (err)
		goto done;

	if (!quiet)
		printf("configuration OK\n");
done:
	if (fd != -1 && fd != STDIN_FILENO && close(fd) == -1 && err == NULL)
		err = got_error_from_errno2("close", configfile);
	free(configfile);
	return err;
}

static void
list_commands(FILE *fp)
{
	size_t i;

	fprintf(fp, "commands:");
	for (i = 0; i < nitems(gotsys_commands); i++) {
		const struct gotsys_cmd *cmd = &gotsys_commands[i];
		fprintf(fp, " %s", cmd->cmd_name);
	}
	fputc('\n', fp);
}

__dead static void
usage(int hflag, int status)
{
	FILE *fp = (status == 0) ? stdout : stderr;

	fprintf(fp, "usage: %s [-hV] command [arg ...]\n", getprogname());
	if (hflag)
		list_commands(fp);
	exit(status);
}

int
main(int argc, char *argv[])
{
	const struct gotsys_cmd *cmd;
	int ch, i;
	int hflag = 0, Vflag = 0;
	static const struct option longopts[] = {
	    { "version", no_argument, NULL, 'V' },
	    { NULL, 0, NULL, 0 }
	};

	setlocale(LC_CTYPE, "");

#ifndef PROFILE
	if (pledge("stdio rpath wpath cpath unix sendfd unveil", NULL) == -1)
		err(1, "pledge");
#endif

	while ((ch = getopt_long(argc, argv, "+hV", longopts, NULL)) != -1) {
		switch (ch) {
		case 'h':
			hflag = 1;
			break;
		case 'V':
			Vflag = 1;
			break;
		default:
			usage(hflag, 1);
			/* NOTREACHED */
		}
	}

	argc -= optind;
	argv += optind;
	optind = 1;
	optreset = 1;

	if (Vflag) {
		got_version_print_str();
		return 0;
	}

	if (argc <= 0)
		usage(hflag, hflag ? 0 : 1);

	for (i = 0; i < nitems(gotsys_commands); i++) {
		const struct got_error *error;

		cmd = &gotsys_commands[i];

		if (strncmp(cmd->cmd_name, argv[0], strlen(argv[0])) != 0)
			continue;

		if (hflag)
			cmd->cmd_usage(0);

		error = cmd->cmd_main(argc, argv);
		if (error) {
			fprintf(stderr, "%s: %s\n", getprogname(), error->msg);
			fflush(stderr);
			return 1;
		}

		return 0;
	}

	fprintf(stderr, "%s: unknown command '%s'\n", getprogname(), argv[0]);
	list_commands(stderr);
	fflush(stderr);
	return 1;
}
